jaeger-core/src/main/java/com/uber/jaeger/Configuration.java

```diff
@@ -263,6 +264,354 @@ public void setStatsFactory(StatsFactory statsFactory) {
+      this.serviceName = serviceName;
+      return this;
+    }
+
+    /**
+     * Sets the {@link SamplerConfiguration} to use. If the sampler type or parameter have been set already, they
+     * will be overridden by the given SamplerConfiguration.
+     *
+     * @see SamplerConfiguration
+     * @param samplerConfiguration the SamplerConfiguration to use
+     * @return the same builder's instance
+     */
+    public Builder withSamplerConfiguration(SamplerConfiguration samplerConfiguration) {
+      this.samplerConfiguration = samplerConfiguration;
+      return this;
+    }
+
+    /**
+     * Sets the {@link ReporterConfiguration} to use
+     * @see ReporterConfiguration
+     * @param reporterConfiguration the ReporterConfiguration to use
+     * @return the same builder's instance
+     */
+    public Builder withReporterConfiguration(ReporterConfiguration reporterConfiguration) {
+      this.reporterConfiguration = reporterConfiguration;
+      return this;
+    }
+
+    /**
+     * Sets the {@link CodecConfiguration} to use
+     * @see CodecConfiguration
+     * @param codecConfiguration the CodecConfiguration to use
+     * @return the same builder's instance
+     */
+    public Builder withCodecConfiguration(CodecConfiguration codecConfiguration) {
+      this.codecConfiguration = codecConfiguration;
+      return this;
+    }
+
+    /**
+     * Sets the {@link StatsFactory} to use. If {@link #withStatsReporter(StatsReporter)} has been called already,
+     * its outcome is overridden by the parameter provided to this method.
+     * @see StatsFactory
+     * @see #withStatsReporter(StatsReporter)
+     * @param statsFactory the StatsFactory to use
+     * @return the same builder's instance
+     */
+    public Builder withStatsFactory(StatsFactory statsFactory) {
+      this.statsFactory = statsFactory;
+      return this;
+    }
+
+    /**
+     * Creates a new {@link StatsFactory} based on the given {@link StatsReporter}. If
+     * {@link #withStatsFactory(StatsFactory)} has been called already, it gets overridden by the {@link StatsFactory}
+     * created by this method.
+     * @see StatsReporter
+     * @see #withStatsFactory(StatsFactory)
+     * @param statsReporter the StatsReporter to use
+     * @return the same builder's instance
+     */
+    public Builder withStatsReporter(StatsReporter statsReporter) {
+      this.statsFactory = new StatsFactoryImpl(statsReporter);
+      return this;
+    }
+
+    /**
+     * Sets the sampler type and parameter to use when constructing a new {@link SamplerConfiguration}. If a
+     * {@link SamplerConfiguration} has been set already, it will be overridden.
+     *
+     * @see SamplerConfiguration#param
+     * @see SamplerConfiguration#type
+     * @param type  the sampler type
+     * @param param the sampler parameter
+     * @return the same builder's instance
+     */
+    public Builder withSamplerTypeAndParam(String type, Number param) {
```

Comment:
There is a `withSamplerConfiguration` up there. This option is to allow simple callers to avoid creating an intermediate `SamplerConfiguration`, making it simpler for a consumer to create a "default" tracer, overriding only these two parameters.

Reply:
I agree with @pavolloffay here. 

I'd rather have these convenience option be owned by `SamplerConfiguration` or something similar. 

I think that the current approach of having to do the following is intuitive enough:

`Tracer.builder("serviceName").withSamplerConfiguration(new SamplerConfiguration(type, param))`.

I feel that the current Builder knows too much about the fields of the classes that its holding, leading to pain when those classes change.

Reply:
> I'd rather have these convenience option be owned by SamplerConfiguration or something similar.

I see those methods as the equivalent of env vars: a consumer setting the two env vars is not required to know about `SamplerConfiguration`. To me, the following example summarizes the advantages (from the readme):

Current: 
```java
Configuration.SenderConfiguration senderConfiguration = new Configuration.SenderConfiguration.Builder()
  .endpoint("https://jaeger-collector:14268/api/traces")
  .build();

ReporterConfiguration reporterConfiguration = new ReporterConfiguration(senderConfiguration.getSender());
Configuration configuration = new Configuration("service-name", SamplerConfiguration.fromEnv(), reporterConfiguration);
io.opentracing.Tracer tracer = configuration.getTracer();
```

Using `withSenderConfiguration`:
```java
Configuration configuration = Configuration
  .builder("Test")
  .withSenderConfiguration(
    new Configuration.SenderConfiguration.Builder()
      .endpoint("https://jaeger-collector-custom:14268/api/traces")
      .build()
  )
  .build();
io.opentracing.Tracer tracer = configuration.getTracer();
```

The way this PR is right now:
```java
Configuration configuration = Configuration.builder("service-name")
  .withCollectorEndpoint("https://jaeger-collector-custom:14268/api/traces")
  .build();
io.opentracing.Tracer tracer = configuration.getTracer();
```

Reply:
Note that the above assumes `withSenderConfiguration` as part of the main builder, and not part of the `ReporterConfiguration` as it would need to be.

Reply:
Thank you for that example. 

>I see those methods as the equivalent of env vars: a consumer setting the two env vars is not required to know about SamplerConfiguration

This is true, the env variables are prefixed with `JAEGER_SAMPLER`, which does give an indication that they are modifying the sampler. 

I agree that your approach provides a simpler API to callers, but I'm concerned that the builder is tightly coupled with parameters it shouldn't concern itself with. 

That being said, perhaps the user benefit outweighs this cost. Ideally, we'd have a nested fluent builder, but I don't see how to do that in Java 6.

Reply:
How about we mark the builder as "experimental" and see how hard it is to maintain in the long run? If it's too much of a burden, we remove the "deep settings" from the main builder.

Reply:
I don't think it's good to mark this as experimental. It's a core configuration struct. 

Another thing, this API shouldn't be aligned with what is exposed via env vars. For example the signature of this method is defined as `public Builder withSamplerTypeAndParam(String type, Number param)` the caller usually does not know string type of the sampler, signature `withSampler(Sampler sampler)` is easier to consume via programmatic API.

Reply:
I think there is some repetition between `TracerBuilder` and `Configuration` classes. Ideally, builders would be tightly coupled with its objects, for example `ReporterBuilder` would be inside the `Reporter` class/package.

This configuration class would just construct these builders based on supplied configuration one being environmental variables.

Reply:
If not a single builder at the top-level `Configuration` object, how about a consumer API that would look like this?

```java
    Configuration
        .builder("the-service")
        .reporter()
          .withAgentHost("agenthost")
          .build()
        .build();
```

`reporter()` returns a `ReporterConfiguration.Builder`, whose `build()` returns a `Configuration.Builder`, allowing the consumer to return back to the original builder.

Another option is to have a `ReporterConfiguration.Builder` that returns a `ReporterConfiguration` on `build()`, allowing for a consumer API to look like this:

```java
    Configuration
        .builder("the-service")
        .withReporterConfiguration(new ReporterConfiguration.Builder()
            .withAgentHost("agenthost")
            .build()
        )
        .build();
```

Reply:
By the way, as a consumer of Jaeger, I still think it's overly complex to require me to use two builders just to override the agent host... Let's remember that we are just a lib being consumed by a real application.

Reply:
The second approach looks better. Although it does not solve the initial concern about adding delegating options to the main conf builder.

Reply:
> Although it does not solve the initial concern about adding delegating options to the main conf builder.

Then I'm really not understanding something... Could you provide an example of how the consumer API would look like?

Reply:
Just without options like this  `withSamplerTypeAndParam(String type, Number param)` and rather accept the builder for `SamplerConfiguration`

Reply:
Discussion moved to https://groups.google.com/d/msgid/jaeger-tracing/1518708828.4245.32.camel%40redhat.com